codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java

```diff
@@ -35,51 +46,78 @@
         /**
          * {@code :method}.
          */
-        METHOD(":method"),
+        METHOD(":method", REQUEST_PSEUDO_HEADER),
 
         /**
          * {@code :scheme}.
          */
-        SCHEME(":scheme"),
+        SCHEME(":scheme", REQUEST_PSEUDO_HEADER),
 
         /**
          * {@code :authority}.
          */
-        AUTHORITY(":authority"),
+        AUTHORITY(":authority", REQUEST_PSEUDO_HEADER),
 
         /**
          * {@code :path}.
          */
-        PATH(":path"),
+        PATH(":path", REQUEST_PSEUDO_HEADER),
 
         /**
          * {@code :status}.
          */
-        STATUS(":status");
+        STATUS(":status", RESPONSE_PSEUDO_HEADER);
+
+        private static char PSEUDO_HEADER_PREFIX = ':';
 
         private final AsciiString value;
-        private static final CharSequenceMap<AsciiString> PSEUDO_HEADERS = new CharSequenceMap<AsciiString>();
+        private final HeaderType type;
+        private static final CharSequenceMap<PseudoHeaderName> PSEUDO_HEADERS = new CharSequenceMap<PseudoHeaderName>();
+
         static {
             for (PseudoHeaderName pseudoHeader : PseudoHeaderName.values()) {
-                PSEUDO_HEADERS.add(pseudoHeader.value(), AsciiString.EMPTY_STRING);
+                PSEUDO_HEADERS.add(pseudoHeader.value(), pseudoHeader);
             }
         }
 
-        PseudoHeaderName(String value) {
+        PseudoHeaderName(String value, HeaderType type) {
             this.value = AsciiString.cached(value);
+            this.type = type;
         }
 
         public AsciiString value() {
             // Return a slice so that the buffer gets its own reader index.
             return value;
         }
 
+        public HeaderType type() {
```

Comment:
does `HeaderType` need to be public? From the perspective of a `PseudoHeaderName` it is only allowed in either `request` or `response`, correct? If that is the case can we just have a method like this:

```java
// true -> the pseudo header can only be used in requests, false -> the pseudo header can only be used in responses
public boolean isRequestOnly() {
}
```

Reply:
`HeaderType` can be `REGULAR_HEADER` , `REQUEST_PSEUDO_HEADER` and `RESPONSE_PSEUDO_HEADER`, we use it to track the current header type in the decoder to help with validation. 

We can make `HeaderType` package private or if we absolutely don't want it, we could try to rely on `Http2Headers.PseudoHeaderName` only but that would make the code more confusing I think.

Reply:
Yeah makes perfect sense, I'll move `HeaderType` to `HPackDecoder` .

Reply:
- [X] Move HeaderType to HPackDecoder